File: Solutions/Python/lru_cache.py
"""
    [MEDIUM]
    146. LRU Cache

    Concepts:
    - hash table
    - linked list

    Stats:
        Runtime | 88 ms     [Beatd 97.31%]
        Memory  | 78.06 MB  [Beats 29.78%]
"""

class LRUCache:

    def __init__(self, capacity: int):
        self.cache = {}     # stores inserted keys and values
        self.capacity = capacity

    def get(self, key: int) -> int:
        if key in self.cache:
            # replace key
            val = self.cache[key]
            del self.cache[key]
            self.cache[key] = val

            return val

        return -1

    def put(self, key: int, value: int) -> None:
        # if key in cache, update value
        if key in self.cache:
            del self.cache[key]
        self.cache[key] = value

        # maintain cache
        if len(self.cache) > self.capacity:
            # delete least recently used
            lru = next(iter(self.cache))
            del self.cache[lru]


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: Solutions/Cpp/lru_cache.cpp
/*
    [MEDIUM]
    146. LRU Cache

    Concepts:
    - design
    - hash table
    - linked list

    Stats:
        Runtime | 46 ms     [Beats 96.50%]
        Memory  | 173.24 MB [Beats 62.89%]
*/

class LRUCache {
    public:
        int capacity;
        std::unordered_map<int, std::list<std::pair<int, int>>::iterator> cache;
        std::list<std::pair<int, int>> dll;
    
        LRUCache(int _capacity) {
            capacity = _capacity;
            cache = {};
            dll = {};
        }
        
        int get(int key) {
            if (cache.contains(key)) {
                dll.splice(dll.begin(), dll, cache[key]);
                return cache[key]->second;
            }
            return -1;
        }
        
        void put(int key, int value) {
            if (cache.contains(key)) {
                dll.splice(dll.begin(), dll, cache[key]);
                cache[key]->second = value;
                return;
            }
    
            dll.push_front({key, value});
            cache.insert({key, dll.begin()});
    
            if (dll.size() > capacity) {
                int lru = dll.back().first;
                dll.pop_back();
                cache.erase(lru);
            }
        }
    };
    
    /**
     * Your LRUCache object will be instantiated and called as such:
     * LRUCache* obj = new LRUCache(capacity);
     * int param_1 = obj->get(key);
     * obj->put(key,value);
     */